### ISSUE TITLE
Asynchronous core normalization

### ISSUE THREAD

**Issue by matthewp** [systemjs/systemjs#483]
This means if you are using a normalize extension you can not use it with plugins. Caused by

https://github.com/systemjs/systemjs/blob/b299e4f46dbf5dcaccdb26c7ea37ac8fd10f1c41/lib/plugins.js#L45

and 

https://github.com/systemjs/systemjs/blob/b299e4f46dbf5dcaccdb26c7ea37ac8fd10f1c41/lib/plugins.js#L53

Why was this change made?

**Comment by guybedford** [comment]
I did put some thought to this actually before release, and one idea was to have normalizeSync be called from within normalize. Then if you want to hook the "synchronous normalize layer" you would actually hook the normalizeSync function itself, which would be called indirectly.

Would something like that help?

**Comment by matthewp** [comment]
First, why is a synchronous normalize needed by the plugin extension?

**Comment by guybedford** [comment]
The configuration normalization into URL space in `System.config` has to be synchronous otherwise configuration calls themselves would be promise-based.

**Comment by matthewp** [comment]
That's a separate matter, why does the plugin extension need to call the synchronous version? The plugin extension is not synchronous.

**Comment by matthewp** [comment]
Ah, I see now, you want System.config to apply to plugin identifiers.  Unfortunately normalize is not synchronous and you should not assume that all normalize extensions can be implemented synchronously (maybe some need to do an import).  Personally I think you should just back portable module names that are user-friendly and this issue doesn't exist, but aside from that you could have a queue of names that are being normalized and block loading until that queue is emptied.

**Comment by matthewp** [comment]
pseudo-code:

```js
System.pendingConfigs = [];

System.config = function(cfg){
  var pendingConfigs = this.pendingConfigs;

  for(name in cfg) {
    pendingConfigs.push(this.normalize(name, "@configuration").then(function(normalizedName){
      // Set your config options here with the fully normalized name.

      // Remove this pendingConfig.
      pendingConfigs.splice(pendingConfigs.indexOf(name), 1));
    });
  }
};

var normalize = System.normalize;
System.normalize = function(name, parentName){
    if(parentName !== "@configuration") {
      // Wait for pending imports
      return Promise.all(pendingImports).then(function(){
        return normalize.apply(System, arguments);
      });
    }
};
```

**Comment by guybedford** [comment]
@matthewp yes that could work, but it seems this is just to solve the issue of allowing an asynchronous normalize right? Normalize is still asynchronous, we've just separated it into a base-level synchronous normalize (applying map, paths, and relative normalization into URLs), and an asynchronous normalize (plugins, conditionals etc etc). Why do you need the npm plugin identifier itself to be synchronously resolved?

**Comment by guybedford** [comment]
@matthewp the npm functionality could be written as an extension, and then it would have full access to async normalization.

**Comment by matthewp** [comment]
it is written as an extension.  

> Why do you need the npm plugin identifier itself to be synchronously resolved?

I don't, I need it to be asynchronously resolved.  It's currently synchronously resolved here: https://github.com/systemjs/systemjs/blob/b299e4f46dbf5dcaccdb26c7ea37ac8fd10f1c41/lib/plugins.js#L53

**Comment by guybedford** [comment]
If it is written as an extension, then surely your extension can impose the necessary asynchronous behaviour in the pipeline? For example, you could pre-normalize the name with the asynchronously resolving part before reaching the plugin system.

**Comment by matthewp** [comment]
Then I've effectively recreated the plugin system if I have to do the work of detecting a plugin and then pre-normalizing both sides.   Now I have to do this for every extension that hooks normalize?

**Comment by guybedford** [comment]
I'm not sure this functionality really is a plugin though? Loading from npm, is more like a "mode" or "resolver" than a plugin, which is more for resources or translations.

You could basically treat this like syntax - `!npm` at the end of a name means doing an async lookup and then normalizing an adjusted name. Let me know if I'm missing something important here though.

**Comment by MajorBreakfast** [comment]
I think the `normalizeSync` method doesn't fit with the rest of the project.

Why not make `System.config()` async. The other parts of the loader that access the config (i.e. normalize hook) then could simply wait until the pending config change is applied. That way the actual usage of `System.config()` must not change. (However, the returned promise can be used if necessary)

**Comment by guybedford** [comment]
So I'd really like it in the new spec if it could work out to have something like:

```javascript
Reflect.loader.resolve('name') // synchronous resolution function
Reflect.loader.hook('resolve', ...); // async resolution hook function
```

It creates a natural separation between the more dynamic resolution behaviours, and more fixed resolution code, is useful to users and is a nice API.

Whether we might get that, I don't know. `resolveSync` works for me in the mean time.

**Comment by matthewp** [comment]
> I'm not sure this functionality really is a plugin though? Loading from npm, is more like a "mode" or "resolver" than a plugin, which is more for resources or translations.

> You could basically treat this like syntax - !npm at the end of a name means doing an async lookup and then normalizing an adjusted name. Let me know if I'm missing something important here though.

I have no idea what that second paragraph means.  Here's a simple example:

```js
import foo from "foo!bar";
```

Here either `foo` or `bar` could be an npm package. Let's say `bar` is.  In 0.16 this would be normalized to `foo!bar@1.0.0#main`.  The part after the bang being the normalized npm name.  Now it is not, because you no longer normalize plugins or plugin resources.

This bug applies to all normalize extensions, not just the npm one used as an example here.  Surely you recognize that this is a huge degradation in functionality.  And for what gain?

**Comment by matthewp** [comment]
> It creates a natural separation between the more dynamic resolution behaviours, and more fixed resolution code, is useful to users and is a nice API.

There's nothing useful about partial normalization.  Either you have the key or you don't.

**Comment by guybedford** [comment]
Ok right I see - the base-level normalization has to be async in this resolver. I wasn't aware you were relying on this contract.

Does this apply to the browser or Node specifically?

**Comment by matthewp** [comment]
I don't know what you're asking or what "base-level normalization" means.

**Comment by guybedford** [comment]
Just summarizing. I assume you're doing the Node lookup here loading the package.json and traversing node_modules with 404s? Do you run this in the browser or just Node?

**Comment by matthewp** [comment]
Both.  There are no 404s for the most part. The only time there are is when code is using the `folderName/index.js` convention and loading it as just `require("folderName")`.  But luckily that coding style isn't as popular for browser libraries.

**Comment by guybedford** [comment]
Right, I finally get it. Yes it's definitely important to keep this use case. Labelled as a bug.

**Comment by MajorBreakfast** [comment]
@guybedford Thanks!

**Comment by guybedford** [comment]
Fix included in https://github.com/systemjs/systemjs/pull/510.

**Comment by guybedford** [comment]
Merged in https://github.com/systemjs/systemjs/commit/8728ead0656966e55ca68e0b62411c17c8d70dd5.

**Comment by guybedford** [comment]
@matthewp unfortunately this has turned out to be a trickier problem - synchronous plugin normalization is actually assumed when bundling to be able to convert bundle names into environment names. We can look at adding support for asynchronous normalization here, but it will involve storing a definition promise, and then carefully checking that all accesses to `loader.defined` first wait on this promise.

It will be a little bit of careful work, but should be relatively straightforward to implement and shouldn't take that long.

If you want to chat further about this feel free to contact me on chat as well.

**Comment by MajorBreakfast** [comment]
@guybedford Thanks for the hard work! Supporting async instead of sync is hard. That's why many lazy programmers go sync and try to ignore that JavaScript has an event loop at its heart. So thanks for doing it right :)

**Comment by matthewp** [comment]
@guybedford I would rethink why you need synchronous normalize in your build.  In my opinion it was a mistake to rush to adopt the new spec when hacks like this were needed to make it work.  What you should probably do is split normalize into 2, one that you call `resolve` that generates the url to use as a key and `normalize` which does what normalize does in the es6 spec, this way you can use normalize to create "environment names" in your build.

**Comment by guybedford** [comment]
@matthewp good point - ideally that would exactly be the way we separate the hook from the user function, but it is yet to be seen how this works out in the spec. For now `System.normalize` is effectively the normalize hook, and `System.normalizeSync` is the sync resolve user function (although not an official API currently).

I've managed to implement the feature based on your suggestion though under these names.

**Comment by guybedford** [comment]
This is now released in 0.18.
